apps/project/serializers.py

```diff
@@ -490,3 +498,106 @@ class ProjectRecentActivitySerializer(serializers.Serializer):
+        )
+        return instance
+
+    def validate_project(self, project):
+        project = get_object_or_404(Project, id=project)
+        if project.is_private:
+            raise serializers.ValidationError("Cannot join private project")
+        if ProjectMembership.objects.filter(project=project, member=self.context['request'].user).exists():
+            raise serializers.ValidationError("Already a member")
+        return project
+
+    def validate_reason(self, reason):
+        if len(reason) <= MIN_LENGTH:
+            raise serializers.ValidationError(gettext("Must be at least %s characters") % MIN_LENGTH)
+        return reason
+
+
+class ProjectAcceptRejectSerializer(serializers.ModelSerializer):
+    id = IntegerIDField(required=False)
+    status = serializers.CharField(required=True)
+    role = serializers.CharField(required=False, allow_null=True, allow_blank=True)
+
+    class Meta:
+        model = ProjectJoinRequest
+        fields = ('id', 'status', 'role')
+
+    @staticmethod
+    def _accept_request(responded_by, join_request, role):
+        if not role or role == 'normal':
+            role = ProjectRole.get_default_role()
+        elif role == 'admin':
+            role = ProjectRole.get_default_admin_role()
+        else:
+            role_qs = ProjectRole.objects.filter(id=role)
+            if not role_qs.exists():
+                raise serializers.ValidationError('Role doesnot exist')
+            role = role_qs.first()
+
+        join_request.status = 'accepted'
+        join_request.responded_by = responded_by
+        join_request.responded_at = timezone.now()
+        join_request.role = role
+        join_request.save()
+
+        ProjectMembership.objects.update_or_create(
+            project=join_request.project,
+            member=join_request.requested_by,
+            defaults={
+                'role': role,
+                'added_by': responded_by,
+            },
+        )
+
+    @staticmethod
+    def _reject_request(responded_by, join_request):
+        join_request.status = 'rejected'
+        join_request.responded_by = responded_by
+        join_request.responded_at = timezone.now()
+        join_request.save()
+
+    def update(self, instance, validated_data):
+        validated_data['project'] = self.context['request'].active_project
+        role = validated_data.pop('role', None)
+        if instance.status in ['accepted', 'rejected']:
+            raise serializers.ValidationError(
+                'This request has already been {}'.format(instance.status)
+            )
+        if validated_data['status'] == 'accepted':
+            ProjectAcceptRejectSerializer._accept_request(self.context['request'].user, instance, role)
+        elif validated_data['status'] == 'rejected':
+            ProjectAcceptRejectSerializer._reject_request(self.context['request'].user, instance)
+        return super().update(instance, validated_data)
```

Comment:
Looks like we can merge this two serializer now. Let's use partial=True, that way create validator isn't triggered in update (if they are provided throw an error if different values are provided)